### ISSUE TITLE
PersonaChat / ConvAI2 - the use of candidate labels in the retrieval setup

### ISSUE THREAD

**Issue by ishalyminov** [facebookresearch/ParlAI#1570]
Hi,

Am I right that IR/ranking models for PersonaChat / ConvAI2 are supposed to retrieve responses from observations' `label_candidates`, both in training and testing?
If not, how do I access the list of all response candidates for these tasks in ParlAI?

**Comment by emilydinan** [comment]
Hi @ishalyminov -- during evaluation the models indeed rank the set of 20 candidates contained in each observation's `label_candidates`. What each model ranks during training is model-dependent -- are you asking about a specific model?

**Comment by ishalyminov** [comment]
Thanks a lot!
I was mainly wondering about retrieval models in general, as my model gets some unreasonably high scores when trained to select from label_candidates (using ParlAI's TrainLoop) - must be some bug then..

**Comment by ishalyminov** [comment]
I'm closing it now and will re-open if I need more info. Thanks again for the answer @emilydinan!

**Comment by ishalyminov** [comment]
I think I've found the problem: the correct label would always come last in the `label_candidates`, so a list-wise ranking model would just always select the last one.
After shuffling `label_candidates` in the `observe()` function, the metrics became more realistic.

Is there a mention of it somewhere in the docs that I've missed?

**Comment by uralik** [comment]
hey @ishalyminov , prolly @emilydinan will give a better answer later.

Since you mention the correct label - I assume you are talking about some test-setting metric evaluation. 
As you see [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/metrics.py#L173-L198) hits metric takes already sorted text_cands from the agent and is iterating from the first one there and computing how far is your correct label in this set and is counting corresponding hits along the way. That is why initially the correct label is the last one - so if you do not do anything with your text_cands -- hits@k should be low.

So if your model selects the last candidate from those 20 - then yes it always will be the correct one.
However - this 20 set is usually used only for validation/metric estimation. If you want to make an actual retrieval model - your candidate set should be *much* larger.

For example, you can check how candidates are scored in transformer agent [here](https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/transformer/transformer.py#L158-L182). It can use either fixed preloaded set of encoded candidates or whatever you want. (you can check candidates argument for TorchRankerAgent)

**Comment by emilydinan** [comment]
Hi @ishalyminov -- indeed, for the  personachat task, the correct answer is always the last label candidate. However, any good ranking model should be invariant to the order of the label_candidates, so metrics shouldn't change if you shuffle the label_candidates. 

As @uralik mentioned, a retrieval model should be capable of ranking large sets of candidates at inference time. For instance, you might try interacting with the KVMemNN model by running this script: projects/convai2/baselines/kvmemnn/interactive.py in which it ranks all of the candidates from the training set.

**Comment by ishalyminov** [comment]
Got it, thanks a lot @uralik @emilydinan!
